examples/selective2d/2d_train.py

```diff
@@ -0,0 +1,354 @@
+
+  return model
+
+def pp(model, pp_device_mesh, args):
+  from pippy.IR import annotate_split_points, PipeSplitWrapper 
+  from pippy import split_into_equal_size
+  from pippy.compile import compile_stage
+  from pippy.microbatch import TensorChunkSpec, sum_reducer
+
+  pp_chunks = args.world_size 
+  pp_groups = pp_device_mesh.get_dim_groups()[0]
+
+  output_chunk_spec = (TensorChunkSpec(0), sum_reducer)
+  stage = compile_stage(model, args.rank, args.world_size, pp_chunks, pp_device_mesh, pp_groups, 
+            example_inputs=[X, Y], 
+            output_chunk_spec=output_chunk_spec)
+  
+  print(f'[Rank{_rank}] {stage.submod.print_readable()}')
+  return model, stage 
+
+def pp_and_tp(model, mesh, args):
+  from pippy.compile import compile_stage
+  from pippy.microbatch import TensorChunkSpec, sum_reducer
+
+  pp_dim, tp_dim = 0, 1
+  pp_rank, tp_rank = args.local_rank // args.tp_size, args.local_rank % args.tp_size
+  pp_groups = mesh.get_dim_groups()[pp_dim]
+
+  # TP
+  tp(model, args.n_layer, mesh, 0, tp_dim)
+
+  X, Y = get_rand(args)
+
+  # PP
+  stage = compile_stage(model, pp_rank, args.world_size, args.pp_size, args.device, pp_groups, 
+            example_inputs=[X, Y],
+            )
+  
+  return model, stage
+
+def even_cut(model, args, pp_size, cut={}):
+  from pippy.IR import annotate_split_points, PipeSplitWrapper 
+  cutpoint = args.n_layer // pp_size
+  for i in range(args.n_layer):
+    name = f'transformer.h.{i}'
+    if i > 0 and i % cutpoint == 0:
+      cut[name] = PipeSplitWrapper.SplitPoint.BEGINNING # or END
+
+  annotate_split_points(model, cut)
+
+def after_ar_cut(model, args, pp_size, cut={}):
+  from pippy.IR import annotate_split_points, PipeSplitWrapper 
+  cutpoint = args.n_layer // pp_size
+  for i in range(args.n_layer):
+    name = f'transformer.h.{i}'
+    if i != args.n_layer - 1 and i % cutpoint == cutpoint - 1:
+      cut[f'{name}.mlp.dropout'] = PipeSplitWrapper.SplitPoint.BEGINNING
+
+  annotate_split_points(model, cut)
+
+def pp_and_tp_fg(model, mesh, args, tp_attn_layers=None, tp_mlp_layers=None, cut_fn=even_cut):
+  from pippy.compile import compile_stage
+  from pippy.microbatch import TensorChunkSpec, sum_reducer
+
+  pp_dim, tp_dim = 0, 1
+  pp_rank, tp_rank = args.local_rank // args.tp_size, args.local_rank % args.tp_size
+  pp_groups = mesh.get_dim_groups()[pp_dim]
+
+  # TP
+  # Apply TP to layers if layer_id is in tp_attn / tp_mlp 
+  tp_attn_layers = list(range(args.n_layer)) if tp_attn_layers is None else tp_attn_layers
+  tp_mlp_layers = list(range(args.n_layer)) if tp_mlp_layers is None else tp_mlp_layers
+  for i in range(args.n_layer):
+    name = f'transformer.h.{i}'
+    att = tp_attention(model, f'{name}.attn', mesh, tp_dim)
+    mlp = tp_mlp(model, f'{name}', mesh, tp_dim)
+
+  X, Y = get_rand(args)
+
+  # PP
+  cut_fn(model, args, args.pp_size)
+  stage = compile_stage(model, pp_rank, args.world_size, args.pp_size, args.device, pp_groups, 
```

Comment:
For my learning purpose, does it work if we first do TP and then call `compile_stage`? So DTensor can already be traced by torch.fx?

Reply:
for now we apply TP first and then PP -->  so Dtensor is traced by torch.fx
if we do PP first and TP, the result does not change but PP changes layer name after tracing 
(e.g., transformer.block.i.attn --> transformer_block_i_attn) so we should change the name that we pass to TP.

Reply:
I see. Good to know that DTensor is traceable by torch.fx. I have a n00b question here, what's the difference between first applying TP vs first applying PP?

Reply:
performance-wise, there is no difference. i think only difference is api. pippy's compile stage breaks higher class (e.g., block/transformer/MLP/Attention) into low-level layers (linear) so we just need to be careful of layer name.